validator/project_validator.go

```diff
@@ -1850,178 +1804,68 @@ func validateTaskSyncCommands(p *model.Project, runLong bool) ValidationErrors {
 	return errs
 }
 
-// validateTVDependsOnTV checks that the task in the given build variant has a
-// dependency on the task in the given build variant.
-func validateTVDependsOnTV(source, target model.TVPair, tvToTaskUnit map[model.TVPair]model.BuildVariantTaskUnit) error {
-	if source == target {
-		return errors.Errorf("task '%s' in build variant '%s' cannot depend on itself",
-			source.TaskName, source.Variant)
-	}
-	visited := map[model.TVPair]bool{}
-	var allTVs []model.TVPair
-	for tv := range tvToTaskUnit {
-		visited[tv] = false
-		allTVs = append(allTVs, tv)
-	}
+// validateTVDependsOnTV checks that the dependent task always has a dependency on the depended on task.
+// The dependedOnTask and every other task along the path must run on all the same requester types as the dependentTask
+// and the dependency on the dependedOnTask must be with a status in statuses, if provided.
+func validateTVDependsOnTV(dependentTask, dependedOnTask model.TVPair, statuses []string, project *model.Project) error {
+	g := project.DependencyGraph()
+	tvTaskUnitMap := tvToTaskUnit(project)
 
-	sourceTask, ok := tvToTaskUnit[source]
-	if !ok {
-		return errors.Errorf("could not find task '%s' in build variant '%s'",
-			source.TaskName, source.Variant)
-	}
+	dependentNode := task.TaskNode{Name: dependentTask.TaskName, Variant: dependentTask.Variant}
+	dependedOnNode := task.TaskNode{Name: dependedOnTask.TaskName, Variant: dependedOnTask.Variant}
+	traversal := func(edge task.DependencyEdge) bool {
```

Comment:
It would be nice to modularize this by defining the  traversal function outside of this function.

Reply:
My aesthetic sense likes it inside since it's so specific to the logic of this function, but if you feel strongly I can move it out.

Reply:
Yeah fair; I think a function comment here would just be helpful saying when the function returns true.